File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-6.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} subRoot
     * @return {boolean}
     */

    
       preOrderTraversal(node) {
        if (node === null) {
            return null;
        }

        let result = node.val.toString();
        result += this.preOrderTraversal(node.left);
        result += this.preOrderTraversal(node.right);

        return result;
    }

    isSubtree(root, subRoot) {
        const fullTree = this.preOrderTraversal(root);
        const subTree = this.preOrderTraversal(subRoot);

        return fullTree.includes(subTree);
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:

    def helper(self, root, k, counter):
        if not root:
            return None

        node = self.helper(root.left, k, counter)

        if node:
            return node

        counter[0] +=1
        
        if counter[0] == k:
            return root

        return self.helper(root.right, k, counter)

    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        counter = [0]
        node = self.helper(root, k, counter)

        return node.val if node else -1